### PR TITLE
Convert messages to struct logs

### LINKED ISSUE #5957: [CT-1267] Convert non-structured messages generated in exceptions.py to structured events
We have a number of logging events in which the string message is generated by external utilities, mostly in various functions in exceptions.py.  Because of this, the actual information used to structure the message is not included in the logging event. These events can be identified by the fact that they contain a "msg" field instead of various structured fields.

Refactor these messages so that more of the formatting code is moved into the logging event code, and so that the events capture the strings used to construct the message.

### PR DESCRIPTION
**Pull request by emmyoop**
resolves #5957 

### Description

Convert msg to be built in struct logs

### Checklist

- [x] I have read [the contributing guide](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md) and understand what's expected of me
- [x] I have signed the [CLA](https://docs.getdbt.com/docs/contributor-license-agreements)
- [x] I have run this code in development and it appears to resolve the stated issue
- [x] This PR includes tests, or tests are not required/relevant for this PR
- [x] I have [opened an issue to add/update docs](https://github.com/dbt-labs/docs.getdbt.com/issues/new/choose), or docs changes are not required/relevant for this PR
- [x] I have run `changie new` to [create a changelog entry](https://github.com/dbt-labs/dbt-core/blob/main/CONTRIBUTING.md#adding-a-changelog-entry)

### CODE REVIEW

File: core/dbt/events/helpers.py
Comment (emmyoop) [inline core/dbt/events/helpers.py:16]: I'm not sure I love what I did here.  These were moved out because of some circular dependency issues.  I'm not sure they even belong in the `/events` dir since they're used in various places.

**Comment by emmyoop** [comment]
Happy to wait on this until after the stuct logging changes are merged in main - can change the target for the PR.

**Review by gshank — CHANGES_REQUESTED** [review]
This is a really good start. But I'd like us to handle the "warn_or_error" calls somewhat differently, and convert as many of the "GeneralWarningMsg" instances to specific events instead. So warn_or_error might look something like:
```
def warn_or_error(event, node):
    if flags.WARN_ERROR:
        from dbt.exceptions import raise_compiler_error

        raise_compiler_error(scrub_secrets(event.info.msg, env_secrets()), node)
    else:
        fire_event(event)
```

We're also going to be doing some work on handling and logging exceptions better too, so that "raise_compiler_error" is also likely to see some changes.

One thing people who consume our logs is asking for is being able to categorize the log events and exceptions, and that GeneralWarningMsg does not help.

File: core/dbt/events/functions.py
Comment (peterallenwebb) [inline core/dbt/events/functions.py:222]: I know this code has been moved and isn't new, but should we perhaps fire the event on both branches? Otherwise, I would guess the exception won't be logged.

File: core/dbt/deprecations.py
Comment (emmyoop) [reply core/dbt/deprecations.py:46]: The description got moved into the event message.  It lives in `types.py` now.

File: core/dbt/events/helpers.py
Comment (gshank) [reply core/dbt/events/helpers.py:16]: Yeah, but I can't think of a great place either. If a superior location emerges, we can move it.

**Review by gshank — APPROVED** [review]
This looks really good!  It's so big it's possible I missed something, but I didn't see any issues.

File: core/dbt/adapters/base/impl.py
Comment (VersusFacit) [inline core/dbt/adapters/base/impl.py:1330]: Nit: this paradigm of stringifying arguments to event types now has the option of adding [this class](https://github.com/dbt-labs/dbt-core/blob/8145eed603266951ce35858f7eef3836012090bd/core/dbt/events/base_types.py#L112) as a subclass to the event class definition.

File: core/dbt/events/base_types.py
Comment (VersusFacit) [inline core/dbt/events/base_types.py:6]: nit: extra whitespace or demanded by flake8?

File: core/dbt/events/functions.py
Comment (VersusFacit) [reply core/dbt/events/functions.py:222]: We comfy with importing the exception down yonder inline rather than at the top? Do we have a style convention here?

File: core/dbt/config/runtime.py
Comment (jtcohen6) [inline core/dbt/config/runtime.py:28]: very cool PR! caught one typo while reading through
```suggestion
UnusedResourceConfigPath
```

File: core/dbt/events/functions.py
Comment (emmyoop) [reply core/dbt/events/functions.py:222]: Doing it here actually prevents a circular dependency.  My next task #5958 should help resolve this to a better solution.

File: core/dbt/adapters/base/impl.py
Comment (emmyoop) [reply core/dbt/adapters/base/impl.py:1330]: That class is super cool.  However, it is causing some issues because protobuf requires us to define the type of each field being passed in in a very narrow scope (types cannot be classes ie Exception).  So `exc` comes in as an `Exception` and then does become a string in the message but that's not enough since proto expects the field itself to be a string already.

File: core/dbt/config/runtime.py
Comment (emmyoop) [reply core/dbt/config/runtime.py]: Good eye!

**Pull request merged**
